### ISSUE TITLE
Add support for global 2FA enforcement

### ISSUE THREAD

**Issue by sourceparkrechnerbetrieb** [the-djmaze/snappymail#349]
We would like to enforce 2FA for every user by default. In Rainloop there was an option within the admin panel to enforce 2FA. 
Our mailserver doesn't expose the ports to the internet, so the 2FA would increase the security for our installation.

**Comment by the-djmaze** [comment]
Although there is still a catch on this issue (first login is never 2FA), i made a some-sort of enforcement.

admin -> extensions -> 2fa cog symbol allows you to enforce
When at login the TOTP code is empty, it will redirect to #/settings/two-factor-auth

This is not a real enforcement, but rather annoying at every login (when you don't set it up).

Feel free to modify the extension/plugin to your needs.

**Comment by sourceparkrechnerbetrieb** [comment]
Nice work, thanks a lot.

Would it be possible to redirect every request to the 2FA settings page as long as the user has not enabled 2FA? At the moment it is still possible to use snappymail as usual when 2FA is not enabled by the user.

**Comment by the-djmaze** [comment]
It might be possible, but that requires more changes in
https://github.com/the-djmaze/snappymail/blob/master/plugins/two-factor-auth/js/TwoFactorAuthLogin.js

Something like:
``` javascript
if (rl.settings.get('TwoFactorMissing') && e.detail && 'Login' !== e.detail.viewModelTemplateID) {
    document.location.hash = '#/settings/two-factor-auth';
}
```

Then `TwoFactorMissing` must be set in index.php AND when activated in the settings:
* can't be disabled
* when set TwoFactorMissing = false

**Comment by the-djmaze** [comment]
Did that, but requires SnappyMail v2.15.1 due to a core change

**Comment by the-djmaze** [comment]
You can test now with v2.15.1

**Comment by winkelement** [comment]
> You can test now with v2.15.1

Works fine for me 😃 👍🏽

**Comment by winkelement** [comment]
I just noticed on every page reload i'm taken back to the 2fa setup page despite having 2fa already set up. I then disable/enable 2fa again to view the inbox but only for as long as i don't reload or logout. Should i create a new issue?

**Comment by the-djmaze** [comment]
> Should i create a new issue?

Not needed, here is ok.

**Comment by sourceparkrechnerbetrieb** [comment]
Sorry for the late response. Great work, but I notice the same behavior as @winkelement. When pressing F5 or Strg-F5 I get back to the 2fa setup page.

**Comment by winkelement** [comment]
@the-djmaze Can you reproduce the bug? Is there anything i can do to help identify the issue?

**Comment by sourceparkrechnerbetrieb** [comment]
With release 2.15.3 it got even worse. With enforced 2FA and logged in having 2FA already enabled, I can only navigate within the settings. Clicking on the "Back" button takes you back to the 2FA settings page.

**Comment by the-djmaze** [comment]
Found and fixed. Released v2.16 of the plugin

**Comment by sourceparkrechnerbetrieb** [comment]
@the-djmaze it works fine now. The ticket can be closed in my opinion.
Thanks for the quick implementation! Good work!
